### ISSUE TITLE
Add package for addons to store plugin independent code

### ISSUE THREAD

**Issue by leonardder** [nvaccess/nvda#14359]
### Is your feature request related to a problem? Please describe.
An addon can contain appModules, brailleDisplayDrivers, synthDrivers, globalPlugins and visionEnhancementProviders. However, there is no way for addons to store code that has to be shared between packages. Imagine a case where a constant has to be defined that has to be available in both an appModule and a globalPlugin. I'm pretty sure @josephsl might know of other cases like this.
In my case, I want to store a base class subclassing driverHandler.Driver that will be a base of both a synthDriver and brailleDisplayDriver.

### Describe the solution you'd like
I'd like to suggest the possibility for addons to provide code for the following NVDA packages:

* comInterfaces: for addons requiring com interface definitions that can be resolved with comTypes automatically (i.e. without doing a magic import trick to make the comInterface available to comtypes)
* utils: package that can hold classes and modules that ought to be shared between plugins in the addon

Cc @josephsl and other addon contributors

**Comment by josephsl** [comment]
Hi,

Other possibilities:

* NVDA objects: one way is to define an NVDA object (preferably an overlay class) that is shared between app modules through a global plugin. In practice, this can be done in a number of ways, such as app modules importing the global plugin when choosing overlay classes and the global plugin's choose overlay classes method restricting some conditional code to a limited set of executables (appModule.appName).
* Conditional startup and code deduplication: while global plugins can be told to not initialize everything at startup (perhaps due to secure mode flag or not meeting Windows release requirement), if the ad-on contains ap modules, it would be tempting to put the same set of code in app modules. The proposed feature request would allow this to happen by letting both global plugins and app modules rely on a function or two defined in a common module.
* Communication between global plugins/drivers/app modules/vision enhancement providers: it might be possible that global plugins may rely on things happening in the app module and vice versa (see an ongoing pull request on add-on files repository on Spotify Communicator). Using a common module and a function or two inside this module will allow easier data exchange and communication between these modules without components importing each other (authors need to make sure things are secure and catch errors).

Add-ons that may benefit from the proposed feature request include Office Desk, StationPlaylist (to some extent although the add-on is no longer actively maintained), and parts of Windows App Essentials (only a thing or two are left in the global plugin portion, and it is possible that the global plugin portion will be gone altogether in a future add-on release).

Thanks.

**Comment by leonardder** [comment]
I don't think we should allow addons to store things in the NVDAObjects package, as that would be very powerful. An addon would be able to provide a replacement of something in NVDAObjects without an explicit addition in an appModule or globalPlugin.
Thinking about this some more, the same can happen with the utils module as well. It might be better to add a new package for this expicitly.

**Comment by feerrenrut** [comment]
I'd prefer that addon code isn't mixed in with the standard namespaces. Instead, a solution that maintained the separation of the add-on. 

When importing shared code from within the addon this should be achieved in a way that avoids namespace collisions: Given an addon called `myAddon` you might import code from a `lib` directory within the add-on with `from myAddon.lib import comInterfaces`

- This would help to prevent add-ons from inadvertently depending on each other.
- We don't want to end up in a situation where the addition of a new file in NVDA could break add-ons because of namespace pollution.

I think this idea can be investigated and prototyped. Though just a note, that we'd generally like to move towards a formal plugin API boundary between NVDA and add-ons, so that the internals of NVDA can be updated without breaking add-ons. A more typical "plugin architecture" would achieve this if you're interested in what that would look like.

**Comment by leonardder** [comment]
> When importing shared code from within the addon this should be achieved in a way that avoids namespace collisions: Given an addon called `myAddon` you might import code from a `lib` directory within the add-on with `from myAddon.lib import comInterfaces`

I need to look into this. It would also open the possibility for something like from myAddon.globalPlugins import plugin

> Though just a note, that we'd generally like to move towards a formal plugin API boundary between NVDA and add-ons, so that the internals of NVDA can be updated without breaking add-ons. A more typical "plugin architecture" would achieve this if you're interested in what that would look like.

Could you elaborate on what this would look like?

**Comment by feerrenrut** [comment]
>Could you elaborate on what this would look like?

I suspect you'll find much better explanations than anything I can write up quickly, the architectural pattern "plugin architecture" has been common for decades. The main principle is to inject the interfaces that the plugin (Add-on code) depends on. The add-on should only have access to that interface, anything else is considered an implementation detail that can't be relied on. Some of the benefits:
- A clear boundary for the supported API for plugins.
- By using multiple interfaces, plugins don't need to be updated when there are changes to an interface they don't use.
- Multiple implementations of the interface can be maintained, keeping compatibility despite changing implementations.

If you're asking how we achieve this, this hasn't been prototyped. But some thoughts on the approach / implementation:
- Starting small, E.G. create a set of interfaces specifically for synthDrivers.
- Gradually work through other use-cases with increasing complexity. Looking at real world add-ons and covering the functionality they require.
- These would need to be distinguishable from current add-ons via the manifest.

Eventually, there is the potential to pave the way for running add-ons in another process:
- This could resolve the concerns for add-ons that rely on 32bit dlls when NVDA moves to 64bit
- Could resolve some security and stability concerns.

**Comment by leonardder** [comment]
Thanks for elaborating. While I dig this idea, I think it is a different story and could best be handled in a different issue.
I've done some research.

1. A structure that allows `from myAddon.lib import comInterfaces` is not that trivial. Sure, it can easily be made possible by adding %userConfig%\addons to sys.path, but that would mean that disabled addons can also be imported.
2. Something like `from addons.myAddon.lib import comInterfaces` has similar problems.
Unfortunately, it seems we can't dynamically add a module per addon to sys.modules (e.g. with importlib.util), because an addon is not a module itself.

**Comment by leonardder** [comment]
I discovered a pretty straight forward method to do this already.
An add-on can call `addonHandler.getCodeAddon()` and then `loadModule` on the add-on to get a module from its directory. I think that fully suffices the use case I'm proposing. It makes sense to document it tough.
it is currently using `pkgutil.ImpImporter`, which is deprecated. It makes sense to rewrite this by using importLib.

**Comment by Adriani90** [comment]
cc: @seanbudd, @gerald-hartig for triaging.

**Comment by gerald-hartig** [comment]
@Adriani90 this was triaged in 2022. Does it need to be reviewed?
